KOGITO-535: Quarkus-Keycloak integration to allow securing the GraphQ…

nmirasch opened the pull request:
…L endpoint

nmirasch commented:
@cristianonicolai @mswiderski here is the first secured point with keycloak integration. The next one would be the graphql UI (login at keycloak for generate and pass the token).
To activate keycloak auth with dev mode add: -Dkeycloak -Dquarkus.kogito.dataindex.auth.enabled=true -Dquarkus.kogito.dataindex.auth.requiredrole=confidential  
To execute the test just adding -Dkeycloak must work. It only execute the keycloak related test. I modified the indexingServiceTest to be able to add the token at given(). Not sure if this provide too much value, more that check that all the routes are served. Just take a look and let me know if I'm on the good way. Any comment, thoughts, ideas are very welcome, Thanks!!

mswiderski commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -30,20 +33,65 @@
+                                            });
+            } else {
+                System.out.println("Auth enabled but No user context found");
+                context.fail(500);
```
assuming this 500 is HTTP response code I'd recommend to return 401 instead

nmirasch replied:
done

mswiderski commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -30,20 +33,65 @@
+                                                }
+                                            });
+            } else {
+                System.out.println("Auth enabled but No user context found");
```
let's move this to logger instead of system out

nmirasch replied:
updated

mswiderski commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -30,20 +33,65 @@
+                                                    if (hasAuthority) {
+                                                        handlerRoutingContext.handle(context);
+                                                    } else {
+                                                        System.out.println("User does not have the authority");
```
let's move it to logger instead of system out

nmirasch replied:
done

mswiderski commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -30,20 +33,65 @@
+                                                        context.fail(403);
+                                                    }
+                                                } else {
+                                                    res.cause().printStackTrace();
```
I guess we should return something here, maybe 500?

nmirasch replied:
done

mswiderski commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -30,20 +33,65 @@
     }
+
+    public void handleWithAuthHandler(Handler<RoutingContext> handlerRoutingContext, RoutingContext context) {
+        if (authEnabled) {
```
is this the only way to have it configured? Can't we do it declaratively like it would be done for JAX-RS... asking as I have no clue if it is possible or not...

nmirasch replied:
@mswiderski you mean something like the @RolesAllowed thing?. I was looking to add something like that here but with no success regarding the auth is done at vertx handlers level, but let me give it another try.

mswiderski replied:
yeah, that what I had in mind

nmirasch commented:
depends on https://github.com/kiegroup/kogito-bom/pull/66

nmirasch commented on data-index/data-index-service/pom.xml:
```diff
@@ -93,6 +93,17 @@
+    <dependency>
+      <groupId>org.testcontainers</groupId>
+      <artifactId>testcontainers</artifactId>
+      <version>1.12.3</version>
```
sure, added in https://github.com/kiegroup/kogito-bom/pull/66

nmirasch commented on data-index/data-index-service/src/main/java/org/kie/kogito/index/vertx/VertxRouterSetup.java:
```diff
@@ -30,20 +33,65 @@
+    Boolean authEnabled;
+
+    @Inject
+    @ConfigProperty(name = "quarkus.kogito.dataindex.auth.requiredrole", defaultValue = "none")
```
updated. I just keep the 'required role' as a kogito conf parameter removing prefix quarkus

nmirasch commented:
@mswiderski Thanks for the review!
@cristianonicolai @mswiderski some updates. I have added the profile keycloak i include the keycloak stuff, just add  -Dquarkus.profile=keycloak

cristianonicolai commented on data-index/data-index-service/src/main/resources/application.properties:
```diff
@@ -33,4 +33,12 @@ mp.messaging.incoming.kogito-usertaskinstances-events.bootstrap.servers=localhos
+#oidc - keycloak
+%keycloak.quarkus.oidc.enabled=true
+%keycloak.quarkus.oidc.auth-server-url=http://localhost:8280/auth/realms/quarkus
+%keycloak.quarkus.oidc.client-id=backend-service
```
```suggestion
%keycloak.quarkus.oidc.client-id=kogito-data-index-service
```

cristianonicolai commented on data-index/data-index-service/src/main/resources/application.properties:
```diff
@@ -33,4 +33,12 @@ mp.messaging.incoming.kogito-usertaskinstances-events.bootstrap.servers=localhos
+quarkus.oidc.enabled=false
+#oidc - keycloak
+%keycloak.quarkus.oidc.enabled=true
+%keycloak.quarkus.oidc.auth-server-url=http://localhost:8280/auth/realms/quarkus
```
```suggestion
%keycloak.quarkus.oidc.auth-server-url=http://localhost:8280/auth/realms/kogito
```

cristianonicolai commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakServer.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public Map<String, String> start() {
+        keycloak = new FixedHostPortGenericContainer("quay.io/keycloak/keycloak")
```
does it require that docker is installed?
quite nice solution :)

mswiderski replied:
yeah, was about to say the same, this is really great approach :)

cristianonicolai commented:
@nmirasch there is also some sonar related issues to take a look but overall is really good :)

cristianonicolai commented:
@nmirasch I was also wondering if this could be pushed to Quakus... if oidc is enabled it could automatically protect the entire application.

mswiderski commented on data-index/data-index-service/src/test/java/org/kie/kogito/index/service/KeycloakServer.java:
```diff
@@ -0,0 +1,33 @@
+package org.kie.kogito.index.service;
```
missing license header

nmirasch replied:
ups, added now :)

mswiderski commented on data-index/data-index-service/src/test/resources/quarkus-realm.json:
```diff
@@ -0,0 +1,2039 @@
+{
+  "id": "11d78bf6-6d10-4484-baba-a1388379d68b",
+  "realm": "quarkus",
```
should be updated to kogito as well when the comment from @cristianonicolai about properties is applied

mswiderski commented on data-index/data-index-service/src/test/resources/application.properties:
```diff
@@ -8,4 +8,10 @@ quarkus.log.category."io.restassured".level=DEBUG
+quarkus.infinispan-client.server-list=localhost:11232
+#keycloak
+quarkus.oidc.enabled=false
+quarkus.oidc.auth-server-url=http://localhost:8281/auth/realms/quarkus
```
same adjustment here, instead of `quarkus` use `kogito`

nmirasch commented:
@cristianonicolai  @mswiderski  Updated the renaming of the quarkus realm and client Id. Changed. Added a new execution at main maven-surefire-plugin enabling the keycloak testing instead of a separated profile. It seems it likes a bit more to sonarcloud :). Now to activate keycloak security just add with that -Dquarkus.profile=keycloak at runtime.
Following @cristianonicolai comment regarding the routes handlers, I changed the routing to activate that authCheck on root route when auth enabled as a first handler keeping the other routes as before.
